### PR TITLE
ENH: Connect fieldmap estimation to preprocessing pipeline of individual DWI runs

### PR DESCRIPTION
**Pull request by josephmje**

### CODE REVIEW

File: dmriprep/workflows/dwi/base.py [hunk 1]
```diff
@@ -6,7 +6,7 @@
 
 from niworkflows.engine.workflows import LiterateWorkflow as Workflow
 from ...interfaces import DerivativesDataSink
-
+from sdcflows.workflows.apply.registration import init_coeff2epi_wf
 
 def init_dwi_preproc_wf(dwi_file):
```
Comment (oesteban) [inline dmriprep/workflows/dwi/base.py:11]:
```suggestion
def init_dwi_preproc_wf(dwi_file, has_fieldmap=False):
```

File: dmriprep/workflows/dwi/base.py [hunk 2]
```diff
@@ -30,6 +30,7 @@ def init_dwi_preproc_wf(dwi_file):
     ----------
     dwi_file : :obj:`os.PathLike`
         One diffusion MRI dataset to be processed.
+    sdc : :bool:
```
Comment (oesteban) [inline dmriprep/workflows/dwi/base.py:33]:
```suggestion
    has_fieldmap : :obj:`bool`
        Build the workflow with a path to register a fieldmap to the DWI.
```

File: dmriprep/workflows/dwi/base.py [hunk 3]
```diff
@@ -77,6 +83,11 @@ def init_dwi_preproc_wf(dwi_file):
                 "dwi_file",
                 "in_bvec",
                 "in_bval",
+                # From fmap
```
Comment (oesteban) [inline dmriprep/workflows/dwi/base.py:86]:
```suggestion
                # From SDCFlows
```

**Review by oesteban — COMMENTED** [review]
Looking good - tiny sill nit pick

File: dmriprep/workflows/dwi/base.py [hunk 4]
```diff
@@ -39,6 +41,10 @@ def init_dwi_preproc_wf(dwi_file):
         File path of the b-vectors
     in_bval
         File path of the b-values
+    fmap_coeff
```
Comment (josephmje) [inline dmriprep/workflows/dwi/base.py:44]:
```suggestion
    fmap
        File path of the fieldmap
    fmap_ref
        File path of the fieldmap reference
    fmap_coeff
```

File: dmriprep/workflows/dwi/base.py [hunk 5]
```diff
@@ -39,6 +41,10 @@ def init_dwi_preproc_wf(dwi_file):
...
+        File path of the fieldmap coefficients
+    fmap_mask
+        File path of the fieldmap mask
```
Comment (josephmje) [inline dmriprep/workflows/dwi/base.py:47]:
```suggestion
        File path of the fieldmap mask
    fmap_id
        The BIDS modality label of the fieldmap being used

```

**Comment by josephmje** [comment]
@oesteban Thanks for the refactor to the base workflow. Reading it now and it looks a lot more intuitive.

**Comment by oesteban** [comment]
@josephmje can you make sure you have pushed anything you were working on and hold off of pushing while I finish the SDCFlows updates? I'll update this branch when SDCFlows is ready.

**Comment by josephmje** [comment]
> @josephmje can you make sure you have pushed anything you were working on and hold off of pushing while I finish the SDCFlows updates? I'll update this branch when SDCFlows is ready.

Yup everything I've been working on has been pushed.

**Comment by oesteban** [comment]
Okay, if tests pass then I think we can go ahead and merge.

It still won't do anything with the fieldmaps because our test dataset with fieldmaps (ds001771) does not have ``IntendedFor`` and therefore, the registration and unwarping is disabled at https://github.com/nipreps/dmriprep/pull/140/files#diff-3f08760032797b48bfababab775f0a5f1852b0c33a7a0915122bc68d1bd52133R90

But I think that's a different issue (probably pertaining to SDCFlows, and namely the problem of what to do when no IntendedFor are given for a potential EPI target). And, as currently implemented, the behavior is similar to fMRIPrep's (which does nothing if IntendedFor is not set).

**Comment by josephmje** [comment]
Should we update the docstring for these workflows so that the fieldmap workflows get visualized? Currently, it refers to `ds000206`, which doesn't have any fieldmaps.

**Comment by oesteban** [comment]
> Should we update the docstring for these workflows so that the fieldmap workflows get visualized? Currently, it refers to `ds000206`, which doesn't have any fieldmaps.

Good luck with that :p

It's going to be a rabbit hole of massive dimensions. Plus, SDCFlows has ~94% coverage. Let's trust that.

**Pull request merged**